rfc/plugins.md

```diff
@@ -0,0 +1,206 @@
+ember-cli-deploy. For example:
+
+```javascript
+module.exports = {
+  name: 'ember-cli-deploy-example-plugin',
+  createDeployPlugin: function(options){
+    return {
+      name: options.name,
+      willDeploy: function:(deployment){
+        // do something during the willDeploy phase of the pipeline
+      },
+      didDeploy: function:(deployment){
+        // do something during the didDeploy phase of the pipeline
+      },
+      // etc, see hooks section for a complete list of methods that
+      // may be implemented by a plugin
+    }
+  }
+}
+```
+
+This approach limits the risk of name conflicts at the top-level of the addon.
+It also allows for the plugin author to consult the addon instance during
+creation of the plugin object to make any contextual decisions necessary.
+Finally, it is agnostic with respect to the type of object the plugin is.
+It may be a POJO, a subclass of CoreObject, or maybe an ES6 class.
+
+The `options` argument passed to `createDeployPlugin` will have a `name`
+property. Usually, the `name` will be the plugin name sans the `ember-cli-deploy-`
+prefix, unless a name has been specified as described in Advanced Plugin
+Configuration below.
+
+### The `deployment` object
+
+For each high-level ember-cli-deploy operation, a `deployment` object is created.
+This object is passed to each hook that is invoked on the plugins. It has a number
+of properties that may be of use to a plugin:
+
+Property | file | info
+--- | --- | ---
+`ui` | - | The ember-cli UI object that can be used to write to stdout.
+`appConfig` | `config/environment.js` | The application configuration.
+`config` | stored in `config/deploy.js` | The configuration portion of `config/deploy.js` for the active environment
+`data` | - | Runtime information about the current operation. Plugins can set properties on this object for later use by themselves or another plugin.
+
+### Async operations in hooks
+
+Hook functions can return a promise to block the deployment pipeline.
+Since most deployment involves some sort of IO it makes senses that most
+plugins will want an async function to complete before continuing to the
+next step. If a plugin does not return a promise, then ember-cli-deploy
+proceeds immediately.
+
+If a promise from any of the plugins is rejected then the deployment
+pipeline will stop and ember-cli-deploy will exit. Returned promises that are
+rejected are treated as unrecoverable errors.
+
+### Configuration
+
+By convention, plugin configuration should be kept in `config/deploy.js` and scoped by
+the plugin's name. e.g. for an `ember-cli-deploy-example` plugin, the configuration would
+go in:
+
+```javascript
+// config/deploy.js
+module.exports = {
+  production: {
+    "example": {
+      bucket: "my-app-production",
+      awsAccessKeyId: process.env.AWS_ACCESS_KEY_ID,
+      awsSecretAccessKey: process.env.AWS_SECRET_ACCESS_KEY
+    }
+  }
+};
+```
```

Comment:
I think the config is upside down. I was talking to Aaron about my ideas at
Ember London and I'm writing a new RFC.


Essentially I think a plugin should do exactly one thing:
- Deploy some files to s3
- Deploy a key-value to redis
- Create an s3 bucket
- Send a hip chat message
Etc

The developer can then set up a config that is an ordered list of tasks. A
plugin that uploads files should take a glob. It shouldn't know which files
it should upload in advance (eg assets). For example if I have a b2b and
b2c version, my images and CSS might differ so I might have two asset
upload tasks with different targets.

Third party plugins are now much easier to integrate - they just get added
to the list of tasks and resolved by name.

If task x fails, tasks 1->x-1 are rolled back. It's up to a plugin
how/whether  it implements the rollback function.

The only unknown here is how to bind input of task x to output of task x-1.
I'd like to support js expressions here.

For ease of setting up a lukemelia deploy, we would have a config blueprint
which sets up the task list for that deploy process. We can have other
blueprints inc third party ones for other common processes.

Thoughts?

Bryan